### PR TITLE
Added clientid, clientsecret and tokenendpoint to SASLMechanismOAuthBearer mechanism

### PR DESCRIPTION
**Pull request by cthtrifork**
See https://github.com/redpanda-data/console/issues/699

### CODE REVIEW

File: backend/pkg/kafka/config_helper.go [hunk 1]
```diff
@@ -92,9 +93,18 @@ func NewKgoConfig(cfg *config.Kafka, logger *zap.Logger, hooks kgo.Hook) ([]kgo.
 
 		// OAuth Bearer
 		if cfg.SASL.Mechanism == config.SASLMechanismOAuthBearer {
-			mechanism := oauth.Auth{
-				Token: cfg.SASL.OAUth.Token,
-			}.AsMechanism()
+			var mechanism sasl.Mechanism
+			if cfg.SASL.OAUth.TokenEndpoint != "" {
```
Comment (cthtrifork) [inline backend/pkg/kafka/config_helper.go:97]: Alternatively we could add another config parameter called `TokenType` with options `ShortLived` and `LongLived` but I personally like this better.

File: backend/pkg/config/kafka_sasl_oauth.go [hunk 2]
```diff
@@ -10,25 +10,70 @@
...
+			return fmt.Errorf("OAuth Bearer client credentials must be set")
+		}
+	} else if c.Token == "" {
 		return fmt.Errorf("OAuth Bearer token must be set")
 	}
 
 	return nil
 }
+
+func (c *KafkaSASLOAuthBearer) AcquireToken(ctx context.Context) (string, error) {
```
Comment (cthtrifork) [inline backend/pkg/config/kafka_sasl_oauth.go:49]: We could also use a standard library for this, but I am not sure what opinion you had on including new libraries

File: backend/pkg/config/kafka_sasl_oauth.go [hunk 3]
```diff
@@ -10,25 +10,75 @@
...
+	"net/http"
 )
 
 // KafkaSASLOAuthBearer is the config struct for the SASL OAuthBearer mechanism
 type KafkaSASLOAuthBearer struct {
-	Token string `yaml:"token"`
+	Token         string `yaml:"token"`
+	ClientID      string `yaml:"clientid"`
+	ClientSecret  string `yaml:"clientsecret"`
+	TokenEndpoint string `yaml:"tokenendpoint"`
```
Comment (weeco) [inline backend/pkg/config/kafka_sasl_oauth.go:26]:
```suggestion
	TokenEndpoint string `yaml:"tokenEndpoint"`
```

File: backend/pkg/config/kafka_sasl_oauth.go [hunk 4]
```diff
@@ -10,25 +10,75 @@
...
 }
 
 // RegisterFlags registers all sensitive Kerberos settings as flag
 func (c *KafkaSASLOAuthBearer) RegisterFlags(f *flag.FlagSet) {
 	f.StringVar(&c.Token, "kafka.sasl.oauth.token", "", "OAuth Bearer Token")
+	f.StringVar(&c.ClientID, "kafka.sasl.oauth.clientid", "", "OAuth Bearer Client Id")
+	f.StringVar(&c.ClientSecret, "kafka.sasl.oauth.clientsecret", "", "OAuth Bearer Client Secret")
+	f.StringVar(&c.TokenEndpoint, "kafka.sasl.oauth.tokenendpoint", "", "OAuth Bearer Token Endpoint")
```
Comment (weeco) [inline backend/pkg/config/kafka_sasl_oauth.go:34]:
```suggestion
	f.StringVar(&c.TokenEndpoint, "kafka.sasl.oauth.token-endpoint", "", "OAuth Bearer Token Endpoint")
```

File: backend/pkg/config/kafka_sasl_oauth.go (hunk 2 above)
Comment (weeco) [reply backend/pkg/config/kafka_sasl_oauth.go:51]: What library do you have in mind? I'm mostly afraid of doing something wrong in the OAuth token process, but the code looks straight forward and given that the authority is the other side which decides whether the token is valid or not, I'm also not that afraid.

File: backend/pkg/kafka/config_helper.go (hunk 1 above)
Comment (weeco) [reply backend/pkg/kafka/config_helper.go:97]: Yes seems reasonable 👍

**Review by weeco — COMMENTED** [review]
Hey @cthtrifork ,
thank you for working on this! Generally LGTM, just have some NITs. Additionally, could you document the new properties in the reference config here: https://github.com/redpanda-data/console/blob/master/docs/config/console.yaml ?

**Comment by cthtrifork** [comment]
@weeco Thanks for the review, as you seem to approve the general approach. I have done some integration testing with a few different OIDC providers and confirmed it now works successfully.

When do you think a release could be a possible?

File: backend/pkg/config/kafka_sasl_oauth.go (hunk 4 above)
Comment (cthtrifork) [reply backend/pkg/config/kafka_sasl_oauth.go:38]: I have reverted your suggestion, as I could not get it to work with ENV variable injection.
 `KAFKA_SASL_OAUTH_TOKEN-ENDPOINT` and `KAFKA_SASL_OAUTH_TOKEN_ENDPOINT` did not work.

Keeping it as `tokenendpoint` allowed me to set `KAFKA_SASL_OAUTH_TOKENENDPOINT=`

Comment (weeco) [reply backend/pkg/config/kafka_sasl_oauth.go:38]: The flags are entirely separate from the env variables. The flag name should not have any impact on the env variable name, it's derived from the struct tags (`yaml`).

**Comment by weeco** [comment]
@cthtrifork , we build a new docker image for each push to master, would that be good enough for you? We have an ongoing release branch (rc for v2.3) but since this is a new feature it wouldn't fit there and we only do one feature release per quarter.

**Comment by cthtrifork** [comment]
@weeco That is perfect - we can use the specific version from a SHA

File: backend/pkg/config/kafka_sasl_oauth.go [hunk 5]
```diff
@@ -10,25 +10,95 @@
...
+	"net/url"
+	"strings"
 )
 
 // KafkaSASLOAuthBearer is the config struct for the SASL OAuthBearer mechanism
 type KafkaSASLOAuthBearer struct {
-	Token string `yaml:"token"`
+	Token         string `yaml:"token"`
+	ClientID      string `yaml:"clientid"`
+	ClientSecret  string `yaml:"clientsecret"`
```
Comment (weeco) [inline backend/pkg/config/kafka_sasl_oauth.go:28]:
```suggestion
	ClientID      string `yaml:"clientId"`
	ClientSecret  string `yaml:"clientSecret"`
```

For consistency, lower camel case

File: docs/config/console.yaml [hunk 6]
```diff
@@ -46,7 +46,11 @@ kafka:
   #     realm:
   #     enableFast: true
   #   oauth:
-  #     token: # can be set via the --kafka.sasl.oauth.token flag as well
+  #     token: # can be set via the --kafka.sasl.oauth.token flag as well. Will be used as a long living token if no token-endpoint is specified
+  #     clientid: # can be set via the --kafka.sasl.oauth.clientid flag as well
+  #     clientsecret: # can be set via the --kafka.sasl.oauth.clientsecret flag as well
+  #     tokenendpoint: # can be set via the --kafka.sasl.oauth.token-endpoint flag as well. Will be used to fetch a short living token if specified
```
Comment (weeco) [inline docs/config/console.yaml:52]:
```suggestion
  #     clientId: # can be set via the --kafka.sasl.oauth.clientid flag as well
  #     clientSecret: # can be set via the --kafka.sasl.oauth.clientsecret flag as well
  #     tokenEndpoint: # can be set via the --kafka.sasl.oauth.token-endpoint flag as well. Will be used to fetch a short living token if specified
```

**Review by weeco — COMMENTED** [review]
Last changes and then we should be good to merge 👍

**Review by weeco — APPROVED** [review]
Thanks for implementing

**Comment by weeco** [comment]
Use one of the latest tags from here to use this feature: https://hub.docker.com/r/redpandadata/console-unstable/tags

**Comment by cthtrifork** [comment]
It is working perfectly on our clusters. Thanks for the collaboration!

> @cthtrifork , we build a new docker image for each push to master, would that be good enough for you? We have an ongoing release branch (rc for v2.3) but since this is a new feature it wouldn't fit there and we only do one feature release per quarter.

We will use this tag until a release is ready. Will the unstable image be available until then?
I would rather not save a copy in our docker registry.

**Comment by weeco** [comment]
Yes, we will never delete any docker tags from the registry unless we believe it would be good for our users (e.g. serious security issues). Feel free to use it until the release. Thank you for reporting back too!

**Pull request merged**